docs/src/references/c8y-configuration-management.md

```diff
@@ -162,4 +164,104 @@ Note that:
   for instance `tedge/configuration_change//etc/tedge/mosquitto-conf/c8y-bridge.conf`.
 * Since the type of configuration file is used as an MQTT topic name, the characters `#` and `+` cannot be used in a type name.
   If such a character is used in a type name (or in the path of a configuration file without explicit type),
-  then the whole plugin configuration `/etc/tedge/c8y/c8y-configuration-plugin.toml` is considered ill-formed.
\ No newline at end of file
+  then the whole plugin configuration `/etc/tedge/c8y/c8y-configuration-plugin.toml` is considered ill-formed.
+
+## Configuration files for child devices
+
+To manage configuration files for child-devices the `c8y_configuration_plugin` supports two aspects:
+1) Allowing to associate a configuration file with a cloud's child-device twin
+2) Allowing to consume/provide a configuration file from/to an external device via network
+
+**Details to Aspect 1: Associating to cloud's child-device twin**
+
+For aspect (1) the plugin provides the field `childid` for all records in the `c8y_configuration_plugin` configuration (reference to section 'Configuration' above). That field is interpreted as unique child-device id and the plugin associates the record's configuration file with corresponding cloud's child-device twin. If not provided the configuration file is associated with cloud's thin-edge device twin.
+
+Example:
+```shell
+$ cat /etc/tedge/c8y/c8y-configuration-plugin.toml
+files = [
+    { path = '/etc/tedge/tedge.toml', type = 'tedge.toml' },              # appears in the cloud on the thin-edge device
+    { path = '/etc/child1/foo.conf', type = 'foo', childid = 'child1'  }  # appears in the cloud on the child-device 'child1'
+  ]
+```
+
+Note that the `c8y_configuration_plugin` does not create any child-device twin in the cloud. Instead the clouds child-device twin must be created upfront. However the plugin takes care to set all capabilities (supported operations, config-types) on that child-device twin in the cloud in the same way as it does for the thin-edge device.
+
+
+**Details to Aspect 2: Consuming/providing configuration files via network**
+
+For aspect (2) there are two proposals as below. Decision has to been taken which proposal to follow.
+
+**Proposal 1:**
+
+The `c8y_configuration_plugin` configuration's record field `path` (reference to section 'Configuration' above) can be prefixed with "mqtt://". Then it's value is treated as MQTT topic structure, where another process/external device is expected to provide/consume the configuration file.
+
+The `c8y_configuration_plugin` expects the process/external device always putting the latest config file as retain message to the MQTT topic `{path}`. Then the plugin consumes that retain message whenever a cloud request comes in.
```

Comment:
If I understand correctly, the *same* topic is used by the child device to send its config and by this plugin to send updates. This is really fragile and will lead to race condition.

We need to distinguish the current configuration declared by the child device from the desired configuration pushed from the cloud. We need to make this distinction whatever the protocol we use (MQTT or HTTP).

Reply:
I agree, that definitely makes sense.

Reply:
> If we plan to use the `MQTT`, hope the size of these files will be within the `MQTT` supported `threshold size`. Else we need to fall back to HTTP.

@PradeepKiruvale , are you refering to C8Y's 16KiB limit on MQTT? Instead that section is about the file-transfer channel between the thin-edge device and the external device. So C8Y is not involved, but mosquitto and the MQTT clients running on thin-edge and the external device.

- mosquitto on the thin-edge device:
allows the MQTT max payload size of 256 MB (if it is not reduce with the mosquitto's configuration 'message_size_limit')

- the MQTT client on the thin-edge device:
'rumqttc' we use in thin-edge has a default max package size of 10 kB (as long as I can see in it's doc). However that can be changed by configuration (see https://docs.rs/rumqttc/0.13.0/rumqttc/struct.MqttOptions.html#method.set_max_packet_size). Honestly I didn't found any upper limit for that configuration in the documentation - so I am not sure if there is one, or if it allows the MQTT standard max-size  of 256 MB.

- the MQTT client on the external device:
that is in hands of the customer, we can never know.

To summarize:

Technically I do not see an issue according to the max MQTT payload size. It allows to transfer tens or hundrets of MB, but a common configuration file would be usually much smaller.

On the other hand I see two other drawbacks:
1) Since that MQTT proposal needs to retain those message, a large fleed of child-device with many small config files could lead to a heavy consumption of RAM memory, cause all these retain message are kept in RAM for the whole lifecycle of the configuration files.

2) Even we expect configuration files have a smaller size than the payload limit of MQTT, other thin-edge features that will also need file-transfer from/to external devices (e.g. Software Management) might not comply to those small file size limits. That means would need to introduce another interfaces for those features.

I added both points from above to the pros/cons list: https://github.com/thin-edge/thin-edge.io/pull/1236#discussion_r927035235